SceneAlgo::history : Turn off monitor for plugs we aren't interested in

danieldresser-ie opened the pull request:
This is a fairly naive implementation of the performance fix we've been talking about ( turns out my weird crashes were just due to a thoughtless copy-and-paste error on my part - the naive implementation appears to work fine ).

I should perhaps add a bit more testing ( either a performance test to show this working, or we might want more correctness tests now that this is getting a bit more complex ) - but it does seem to work fine.  It takes the time for a history() call in my production test from 290 seconds to 120 seconds.

The specific way things are set up could potentially use improvement - there are probably a lot of other options for where we store the `disableMonitor` scope ... it's a shame to be taking up an extra pointer in the CapturedProcess struct when we're not using any of the regular pointers at the same time.

johnhaddon commented on src/GafferScene/SceneAlgo.cpp:
```diff
@@ -369,6 +369,7 @@ struct CapturedProcess
 	ConstPlugPtr plug;
 	ConstPlugPtr destinationPlug;
 	ContextPtr context;
+	Monitor::Scope* disableMonitor;
```
Why not `unique_ptr`?

johnhaddon commented on src/GafferScene/SceneAlgo.cpp:
```diff
@@ -403,33 +404,25 @@ class CapturingMonitor : public Monitor
-			if( !p->parent<ScenePlug>() || p->getName() != m_scenePlugChildName )
+
+			CapturedProcess::Ptr capturedProcess( new CapturedProcess );
+			if( p->parent() && ( !p->parent<ScenePlug>() || p->getName() != m_scenePlugChildName ) )
```
What's the extra `p->parent()` check about? I can't see anywhere we're vulnerable to it being null, and I'd have thought it's impossible to start a process for a parent-less plug anyway.

johnhaddon commented on src/GafferScene/SceneAlgo.cpp:
```diff
@@ -403,33 +404,25 @@ class CapturingMonitor : public Monitor
-				m_processMap.insert( ProcessMap::value_type( process, nullptr ) );
-				return;
+				// This isn't a plug we're interested in.  Set a null plug pointer so we won't report
+				// this result, and create a Scope to disable the monitor so that downstream computations
```
downstream -> upstream

johnhaddon commented on src/GafferScene/SceneAlgo.cpp:
```diff
@@ -403,33 +404,25 @@ class CapturingMonitor : public Monitor
+				// This isn't a plug we're interested in.  Set a null plug pointer so we won't report
+				// this result, and create a Scope to disable the monitor so that downstream computations
+				// can go faster without needing to access the monitor
+				capturedProcess->plug = nullptr;
```
This is redundant - `PlugPtr` implicitly constructs to `nullptr` anyway.

johnhaddon replied:
It looks like we're still parenting this CapturedProcess to the parent CapturedProcess (L434 below) so that it gets visited by `historyWalk()`. It seems like we could just not parent it and then we can remove the duplicate "is it a plug we care about?" checks from `historyWalk()`?

johnhaddon commented on src/GafferScene/SceneAlgo.cpp:
```diff
@@ -403,33 +404,25 @@ class CapturingMonitor : public Monitor
+				capturedProcess->plug = p;
+				capturedProcess->destinationPlug = process->destinationPlug();
+				capturedProcess->context = new Context( *process->context(), /* omitCanceller = */ true );
+				capturedProcess->disableMonitor = nullptr;
```
If we used `unique_ptr`, this would be redundant.

johnhaddon commented on src/GafferScene/SceneAlgo.cpp:
```diff
@@ -475,6 +485,16 @@ SceneAlgo::History::Ptr historyWalk( const CapturedProcess *process, InternedStr
+	if( !process->plug )
+	{
+		// CapturedProcess entries with a null plug are just sentinels for plugs that we aren't interested in
+		// TODO - would it be better for sentinels to not store themselves in parent->children()?
+		// m_processMap doesn't store smart pointers with ownership, but we know that we are going to create them
+		// in processStarted and finish with them in processFinished, so we could just do manual memory management
+		// with new/delete
```
Ah, I see why you're storing them now - something has to keep them alive. It does seem wrong for CapturingMonitor to capture things we're not interested in though, and then for us to repeat the filtering code later.

Can we store `variant<CapturedProcess *, MonitorScope>` in the map instead? I think that could be quite a lot cleaner for a couple of reasons :

- The data structure enforces the "either capture, or disable future captures" behaviour we are trying to implement.
- `processFinished()` can go back to being a simple `erase()`.
- We can remove all the manual memory management.
- `historyWalk()` won't have to do any filtering at all (we can remove the previous filtering even).

johnhaddon reviewed (changes requested):
Thanks Daniel - this is a worthy improvement! You've given the before/after timings, but I wonder if you've timed the same hash without the monitor at all? Would be interesting to see how close to zero-overhead we are now.

danieldresser-ie commented:
I think I've addressed all your feedback - most of the code has changed, so I won't reference specific comments.

The test does not show as extreme a change as I'm getting with the production scene, but it takes 5 seconds before #4536, 2 seconds after #4536, and 1 second after this PR, so it is adequate to show a difference ( getting into the really bad scaling might take more than 1 second ).

Oh, and in reply to your earlier question about comparing to a simple attributeHash - in the production case I've been testing, the difference is less than the margin of error ( the margin of error is pretty high though, around 5%.  But definitely no major difference left ).
